File: 2349.design-a-number-container-system.py
# @leet start
from collections import defaultdict
from heapq import heappop, heappush

Heap = list[int]


class NumberContainers:
    # lazy version of the C++ version, which only changes ordering on find
    # SC O(n)

    def __init__(self):
        # reminder that python heaps are by default minheap
        self.num_to_heap: defaultdict[int, Heap] = defaultdict(list)
        self.containers: dict[int, int] = dict()

    def change(self, index: int, number: int) -> None:
        self.containers[index] = number
        heappush(self.num_to_heap[number], index)

    def find(self, number: int) -> int:
        # TC amortised O(logn) of 1 find operation over k change operations
        while self.num_to_heap[number]:
            cont = self.num_to_heap[number][0]
            if self.containers[cont] == number:
                return cont
            heappop(self.num_to_heap[number])

        return -1


# Your NumberContainers object will be instantiated and called as such:
# obj = NumberContainers()
# obj.change(index,number)
# param_2 = obj.find(number)


# Your NumberContainers object will be instantiated and called as such:
# obj = NumberContainers()
# obj.change(index,number)
# param_2 = obj.find(number)
# @leet end

File: 2349.design-a-number-container-system.cpp
// @leet start
#include <set>
#include <unordered_map>

using std::set;
using std::unordered_map;

class NumberContainers {
  private:
    // number to sorted_list of container_ids
    unordered_map<int, set<int>> ordered_lists;
    unordered_map<int, int> containers;

  public:
    NumberContainers() {}

    void change(int index, int number) {
        auto it = containers.find(index);
        if (containers.contains(index)) {
            auto& prev_num = containers[index];
            ordered_lists[prev_num].erase(index);
        }
        containers[index] = number;
        ordered_lists[number].insert(index);
    }

    int find(int number) {
        if (!ordered_lists[number].empty()) {
            // O(logn) of popping treeset to get the smallest index
            return *ordered_lists[number].begin();
        }
        return -1;
    }
};

/**
 * Your NumberContainers object will be instantiated and called as such:
 * NumberContainers* obj = new NumberContainers();
 * obj->change(index,number);
 * int param_2 = obj->find(number);
 */

/**
 * Your NumberContainers object will be instantiated and called as such:
 * NumberContainers* obj = new NumberContainers();
 * obj->change(index,number);
 * int param_2 = obj->find(number);
 */
// @leet end
